File: horseshoe.m
function[betaout]=horseshoe(y,X,BURNIN,MCMC,thin,scl_ub,scl_lb,phasein,a0,b0,BetaTrue)
% Function to impelement Horseshoe shrinkage prior (http://faculty.chicagobooth.edu/nicholas.polson/research/papers/Horse.pdf)
% in Bayesian Linear Regression. %%
% Based on code by Antik Chakraborty (antik@stat.tamu.edu) and Anirban Bhattacharya (anirbanb@stat.tamu.edu)
% Modified by James Johndrow (johndrow@stanford.edu)
% Modified further by Akihiko Nishimura

% Model: y=X\beta+\epslion, \epsilon \sim N(0,\sigma^2) %%
%        \beta_j \sim N(0,\sigma^2 \lambda_j^2 \tau^2) %%
%        \lambda_j \sim Half-Cauchy(0,1), \tau \sim Half-Cauchy (0,1) %%
%        \pi(\sigma^2) \sim 1/\sigma^2 %%


% This function implements the algorithm proposed in "Scalable MCMC for
% Bayes Shrinkage Priors" by Johndrow and Orenstein (2017). 
% The global local scale parameters are updated via a Slice sampling scheme given in the online supplement 
% of "The Bayesian Bridge" by Polson et. al. (2011). Setting ab = true
% implements the algorithm of Bhattacharya et al. Setting ab=false
% implements the algorith of Johndrow and Orenstein, which uses a block
% update for \tau, \sigma^2, \beta



% Input: y=response, a n*1 vector %%
%        X=matrix of covariates, dimension n*p %%
%        BURNIN= number of burnin MCMC samples %%
%        MCMC= number of posterior draws to be saved %%
%        thin= thinning parameter of the chain %%
%        scl_ub=upper bound on scale for MH proposals
%        scl_lb=lower bound on scale for MH proposals (usually make these
%        equal; 0.8 a good default)
%        phasin=number of iterations over which to transition between upper
%        and lower bound on MH proposals; usually make this 1 and just make
%        scl_ub=scl_lb; only use for particularly challenging cases
%        SAVE_SAMPLES=binary; whether to save samples
%        ab=whether to run the algorithm of Bhattachaya et al.
%        trunc=whether to use the numeric truncations of Bhattacharya et al
%        a0=parameter of gamma prior for sigma2
%        b0=second parameter of gamma prior for sigma2
%        BetaTrue=true beta (for simulations)
%        disp_int=how often to produce output
%        plotting=whether to make plots
%        corX=whether simulations were performed with correlated design



% Output: 
%         pMean= posterior mean of Beta, a p by 1 vector%%
%         pMeadian=posterior median of Beta, a p by 1 vector %%
%         pLambda=posterior mean of local scale parameters, a p by 1 vector %%
%         pSigma=posterior mean of Error variance %% 
%         betaout=posterior samples of beta %%


tic;
N=BURNIN+MCMC;
effsamp=(N-BURNIN)/thin;
[n,p]=size(X);

% paramters %
Beta=ones(p,1); 
lambda=ones(p,1);
tau=1; sigma_sq=1;

% output %
betaout=zeros(50,effsamp);
lambdaout=zeros(50,effsamp);
etaout = zeros(50,effsamp);
tauout=zeros(effsamp,1);
xiout = zeros(MCMC+BURNIN,1);
sigmaSqout=zeros(effsamp,1);
l1out = zeros(MCMC+BURNIN,1);
pexpout = zeros(MCMC+BURNIN,1);
ACC = zeros(MCMC+BURNIN,1);

% matrices %
I_n=eye(n); 
l=ones(n,1);

Xi = tau^(-2);
Eta = lambda.^(-2);

% start Gibbs sampling %
for i=1:N  
    
    % update tau %
    if i>0
        LX=bsxfun(@times,(lambda.^2),X');
        XLX = X*LX;

        Eta = lambda.^(-2);
        if i<phasein
           std_MH = (scl_ub.*(phasein-i)+scl_lb.*i)./phasein; 
        else
           std_MH = scl_lb; 
        end
        prop_Xi = exp(normrnd(log(Xi),std_MH));
        lrat_prop = lmh_ratio(XLX,y,prop_Xi,I_n,n,a0,b0);
        lrat_curr = lmh_ratio(XLX,y,Xi,I_n,n,a0,b0);
        log_acc_rat = (lrat_prop-lrat_curr)+(log(prop_Xi)-log(Xi));

        ACC(i) = (rand < exp(log_acc_rat));
        if ACC(i) % if accepted, update
            Xi = prop_Xi;
        end
        tau = 1./sqrt(Xi);
    end
    
    % update sigma_sq marginal of beta %
    M = I_n + (1./Xi).*XLX;
    xtmp = M\y;
    ssr = y'*xtmp;
    sigma_sq = 1/gamrnd((n+a0)/2,2/(ssr+b0));
    
    % Alternative update of sigma_sq conditional on beta
    %{
        if trunc
            E_1=max((y-X*Beta)'*(y-X*Beta),(1e-10)); % for numerical stability
            E_2=max(sum(Beta.^2./((tau*lambda)).^2),(1e-10));
        else
            E_1 = (y-X*Beta)'*(y-X*Beta); E_2 = sum(Beta.^2./((tau*lambda)).^2);
        end
        
        sigma_sq=1/gamrnd((n+p+a0)/2,2/(b0+E_1+E_2));
    %} 

    % update beta %
    U = (1./Xi).*LX;
    u=normrnd(0,tau*lambda);
    v=X*u+normrnd(0,l);
    v_star=(M)\((y./sqrt(sigma_sq))-v);
    Beta=sqrt(sigma_sq)*(u+U*v_star);
    
    % update lambda_j's in a block using slice sampling %
    u = unifrnd(0, 1./(Eta+1));
    gamma_rate = (Beta.^2) .* Xi ./ (2.*sigma_sq);
    Eta = gen_truncated_exp(gamma_rate, (1-u)./u);
    if any(Eta<=0)
       disp([num2str(sum(Eta<=0)) ' Eta underflowed, replacing = machine epsilon']);
       Eta(Eta<=0) = eps;
    end
    lambda = 1./sqrt(Eta);
    
    % (theoretically) equivalent way to sample lambda_j's, but supposedly
    % not as numerically stable.
    %{
    eta = 1./(lambda.^2); 
        upsi = unifrnd(0,1./(1+eta));
        tempps = Beta.^2/(2*sigma_sq*tau^2); 
        ub = (1-upsi)./upsi;

        % now sample eta from exp(tempv) truncated between 0 & upsi/(1-upsi)
        Fub = 1 - exp(-tempps.*ub); % exp cdf at ub 
        Fub(Fub < (1e-4)) = 1e-4;  % for numerical stability
        up = unifrnd(0,Fub); 
        eta = -log(1-up)./tempps; 
        lambda = 1./sqrt(eta);
        Eta = eta;
    %}
    
    per_expl = 1-sqrt(sum((Beta-BetaTrue).^2))./sqrt(sum(BetaTrue.^2));
    L1_loss = 1-sum(abs(Beta-BetaTrue))./sum(abs(BetaTrue));
    
    if i > BURNIN && mod(i, thin)== 0
        betaout(:,(i-BURNIN)/thin) = Beta(1:50);
        lambdaout(:,(i-BURNIN)/thin) = lambda(1:50);
        etaout(:,(i-BURNIN)/thin) = Eta(1:50);
        tauout((i-BURNIN)/thin)=tau;
        xiout(i) = Xi;
        sigmaSqout((i-BURNIN)/thin)=sigma_sq;
        l1out(i) = L1_loss;
        pexpout(i) = per_expl;
    end
end

t=toc;
fprintf('Execution time of %d Gibbs iteration with (n,p)=(%d,%d)is %f seconds. \n',N,n,p,t)

end



function lr = lmh_ratio(XLX,y,Xi,I_n,n,a0,b0)
    % marginal of beta, sigma2
    M = I_n + (1./Xi).*XLX;
    x = M\y;
    ssr = y'*x+b0;
    try
        cM = chol(M); 
        ldetM = 2*sum(log(diag(cM)));
        ll = -.5.*ldetM - ((n+a0)/2).*log(ssr);
        lpr = -log(sqrt(Xi).*(1+Xi));
        lr = ll+lpr;
    catch
        lr = -Inf; warning('proposal was rejected because I+XDX was not positive-definite');
    end
end

function x = gen_truncated_exp(mn,trunc_point)
    r = mn.*trunc_point;
    sml = abs(r)<eps;
    x = zeros(max(length(mn),length(trunc_point)),1);
    tmp = zeros(max(length(mn),length(trunc_point)),1);
    
    if any(sml)
        tmp(sml) = expm1(-r(sml)).*rand(length(mn(sml)),1);
    end
    tmp(~sml) = (exp(-r(~sml))-1).*rand(length(mn(~sml)),1);
    
    sml = abs(tmp)<eps;
    if any(sml)
       x(sml) = -log1p(tmp(sml))./mn(sml); 
    end
    x(~sml) = -log(1+tmp(~sml))./mn(~sml);
    
end

File: horseshoe.py
import numpy as np
import scipy as sp
import scipy.linalg
import math
import warnings
from pypolyagamma import PyPolyaGamma
pg = PyPolyaGamma()

def grad_marginal(sampler, log_tau, lam0, n_burnin, n_samples):
    """
    Returns a Monte Carlo estimate of the gradient of log marginal
    distribution of 'tau'.

    Params
    ------
      sampler(tau, lam0, n_burnin, n_samples): callable
          Returns the samples of 'beta' and 'lam' drawn from the posterior
          conditional distribution.
      tau: float
          The value at which the gradient will be evaluated.
      lam0: vector
          The initial value of local shrinkage parameters for the Gibbs
          sampler.
    """

    # TODO: Update the code to accommodate other values of horseshoe
    # hyper-parameters other than all 1's.

    tau = math.exp(log_tau)
    samples = sampler(tau, lam0, n_burnin, n_samples)
    p = np.size(samples['beta'], 0)

    # Contributions from the likelihood \pi(\beta | \lam, \tau)
    sq_norm_samples = np.sum((samples['beta'] / samples['lambda']) ** 2, 0)
    grad_samples = sq_norm_samples / tau ** 2 - p
    grad = np.mean(grad_samples)
    cov_grad = np.var(grad_samples)
    hess = - 2 / tau ** 2 * np.mean(sq_norm_samples) + cov_grad

    # Contributions from the half-Cauchy prior on 'tau'.
    grad += (1 - tau ** 2) / (1 + tau ** 2)
    hess += - 4 * tau ** 2 / (1 + tau ** 2) ** 2

    # Return a sample of lam to feed into the next iteration.
    lam = samples['lambda'][:, -1]

    return grad, cov_grad, hess, lam


def mcem(sampler, log_tau, lam0, n_burnin, n_samples, include_prior=True):
    """
    Update tau via Monte Carlo EM.

    Params
    ------
      sampler(tau, lam0, n_burnin, n_samples): callable
          Returns the samples of 'beta' and 'lam' drawn from the posterior
          conditional distribution.
      tau: float
          The value with respect to which the incomplete log-likelihood is
          computed.
      lam0: vector
          The initial value of local shrinkage parameters for the Gibbs
          sampler.
      include_prior: bool
          If False, the algorithm tries to maximize only the marginal
          likelihood ignoring the half-Cauchy prior.
    """

    # TODO: Update the code to accommodate other values of horseshoe
    # hyper-parameters other than all 1's.

    tau = math.exp(log_tau)
    samples = sampler(tau, lam0, n_burnin, n_samples)
    p = np.size(samples['beta'], 0)
    sq_norm_samples = np.sum((samples['beta'] / samples['lambda']) ** 2, 0)

    if include_prior:
        tau = math.sqrt(np.mean(sq_norm_samples) / p)
    else:
        a = 1 + p
        b = - (1 - p + np.mean(sq_norm_samples))
        c = - np.mean(sq_norm_samples)
        tau_sq = (- b + math.sqrt(b ** 2 - 4 * a * c)) / 2 / a
        tau = math.sqrt(tau_sq)

    log_tau = math.log(tau)
    lam = samples['lambda'][:, -1]
    return log_tau, lam


def gibbs(y, X, n_burnin, n_post_burnin, thin, tau_fixed=False,
          tau0=None, lam0=None, link='gaussian'):
    """
    MCMC implementation for Bayesian Lasso. Note that the names of 'lambda'
    (local parameter here) and 'tau' (global parameter here) are often switched
    in the literature.

    Model: y = X\beta + \epslion, \epsilon \sim N(0, \sigma^2) 
           \beta_j \sim N(0, \sigma^2 \lambda_j^2 \tau^2)
           \lambda_j \sim Half-Cauchy(0, 1),
           \tau \sim Half-Cauchy (0, 1), 
           \pi(\sigma^2) \sim 1 / \sigma^2


    Input: y = response, a n * 1 vector
           X = matrix of covariates, dimension n * p
           n_burnin = number of burnin MCMC samples
           n_post_burnin = number of posterior draws to be saved
           thin = thinning parameter of the chain
           fixed_penalty = if true, the penalty parameter will not be updated.
           tau0 = the initial value for MCMC
    """

    n_iter = n_burnin + n_post_burnin
    n_sample = math.ceil(n_post_burnin / thin)  # Number of samples to keep
    n, p = np.shape(X)
    if link == 'logit':
        n_trial = np.ones(n)
        kappa = y - n_trial / 2
        omega = n_trial / 2

    # Hyper-parameters
    df_local = 1
    df_global = 1
    global_scale = 1 # scale of the t-distribution prior on 'tau'

    # Initial state of the Markov chain
    beta = np.zeros(p)
    sigma_sq = 1
    if lam0 is not None:
        lam = lam0
    else:
        lam = np.ones(p)
    nu = np.random.gamma((df_local + 1) / 2, 1, size=p) \
         / (1 + df_local / lam ** 2)
        # an auxiliary parameter for sampling 'lam'
    if tau0 is not None:
        tau = tau0
    else:
        tau = 1
    xi = np.random.gamma((df_global + 1) / 2, 1) \
         / (1 + global_scale ** 2 * df_global / tau ** 2)
        # an auxiliary parameter for sampling 'tau'

    # Pre-allocate
    samples = {
        'beta': np.zeros((p, n_sample)),
        'lambda': np.zeros((p, n_sample)),
        'tau': np.zeros(n_sample)
    }
    if link == 'gaussian':
        samples['sigma_sq'] = np.zeros(n_sample)

    # Start Gibbs sampling
    for i in range(n_iter):

        # Update beta and related parameters.
        if link == 'gaussian':
            D = (tau * lam) ** 2 / sigma_sq
            beta = math.sqrt(sigma_sq) \
                   * generate_gaussian(y / math.sqrt(sigma_sq), X, D)
            resid = y - np.dot(X, beta)
            scale = np.sum(resid ** 2) / 2
            sigma_sq = scale / np.random.gamma(n / 2, 1)
        elif link == 'logit':
            pg.pgdrawv(n_trial, np.dot(X, beta), omega)
            D = (tau * lam) ** 2
            beta = generate_gaussian(
                kappa / np.sqrt(omega), np.sqrt(omega)[:, np.newaxis] * X, D)

        else:
            raise NotImplementedError(
                'The specified link function is not supported.')

        # Update local shrinkage parameters via parameter expansion
        # TODO: this data augmentation strategy seems to not-so-good mixing
        # property. Run several iterations or replace with slice sampler.
        scale = df_local * nu + beta ** 2 / 2 / tau ** 2
        lam_sq = scale / np.random.gamma((df_local + 1) / 2, 1, size=p)
        nu = np.random.gamma((df_local + 1) / 2, 1, size=p) \
             / (1 + df_local / lam_sq)
        lam = np.sqrt(lam_sq)

        # Update the global shrinkage parameter
        if not tau_fixed:
            scale = global_scale ** 2 * df_global * xi \
                    + np.sum((beta / lam) ** 2) / 2  # inverse-gamma scale
            tau_sq = scale / np.random.gamma((df_global + p) / 2, 1)
            xi = np.random.gamma((df_global + 1) / 2, 1) \
                 / (1 + global_scale ** 2 * df_global / tau_sq)
            tau = math.sqrt(tau_sq)

        if i >= n_burnin and i % thin == 0:
            index = math.floor((i - n_burnin) / thin)
            samples['beta'][:, index] = beta
            samples['lambda'][:, index] = lam
            samples['tau'][index] = tau
            if link == 'gaussian':
                samples['sigma_sq'][index] = sigma_sq

    return samples


def generate_gaussian(y, X, D, A=None, is_chol=False):
    """
    Generate a multi-variate Gaussian with the mean mu and covariance Sigma of the form
       Sigma = (X'X + D^{-1})^{-1}, mu = Sigma X' y
    where D is assumed to be diagonal.

    Param:
    ------
        A : matrix
            Equals the matrix (XDX' + I) or, if is_chol == true, its cholesky decomposition
        D : vector
    """

    n, p = np.shape(X)
    if n > p:
        Phi = np.dot(X.T, X) + np.diag(D ** -1)
        Phi_chol = sp.linalg.cholesky(Phi)
        mu = sp.linalg.cho_solve((Phi_chol, False), np.dot(X.T, y))
        x = mu + sp.linalg.solve_triangular(Phi_chol, np.random.randn(p),
                                            lower=False)
    else:
        if A is None:
            A = np.dot(X, D[:, np.newaxis] * X.T) + np.eye(n)
        u = np.sqrt(D) * np.random.randn(p)
        v = np.dot(X, u) + np.random.randn(n)
        if is_chol:
            w = sp.linalg.cho_solve((A, False), y - v)
        else:
            w = sp.linalg.solve(A, y - v, sym_pos=True)
        x = u + D * np.dot(X.T, w)
    return x
